components/callbacks/fx.go

```diff
@@ -39,11 +46,32 @@ var Module = fx.Module(
 	fx.Invoke(RegisterExecutor),
 )
 
-func HTTPCallerProviderProvider() HTTPCallerProvider {
+func HTTPCallerProviderProvider(
+	clusterMetadata cluster.Metadata,
+	httpClientCache *cluster.FrontendHTTPClientCache,
+	logger log.Logger,
+) HTTPCallerProvider {
 	m := collection.NewOnceMap(func(queues.NamespaceIDAndDestination) HTTPCaller {
-		// In the future, we'll want to inject headers and certs here.
-		client := &http.Client{}
-		return client.Do
+		return func(r *http.Request) (*http.Response, error) {
+			client := &http.Client{}
+			curClusterName := clusterMetadata.GetCurrentClusterName()
+			clusterInfo, ok := clusterMetadata.GetAllClusterInfo()[curClusterName]
+			if !ok {
+				logger.Warn("HTTPCallerProviderProvider unable to get ClusterInformation for current cluster. Using default HTTP client.", tag.SourceCluster(curClusterName))
+				return client.Do(r)
+			}
+			if r.Header.Get(CallbackSourceHeader) == clusterInfo.ClusterID {
+				internalClient, err := httpClientCache.Get(curClusterName)
+				if err != nil {
+					logger.Warn("HTTPCallerProviderProvider unable to get FrontendHTTPClient for current cluster. Using default HTTP client.", tag.SourceCluster(curClusterName))
+					return client.Do(r)
+				}
+				r.URL.Scheme = internalClient.Scheme()
+				r.URL.Host = internalClient.Host()
```

Comment:
You'll want to also modify `r.Host`.

Reply:
For client requests `r.Host` is just used to populate the host header and if it is unset, it will use the value from the URL. I can do it just for the sake of completeness, but it's not really necessary.